implementations/rust/ockam/ockam_app_lib/src/incoming_services/state.rs

```diff
@@ -0,0 +1,365 @@
+        self.services.iter_mut().find(|s| s.id == id)
+    }
+}
+
+impl ModelState {
+    pub(crate) fn upsert_incoming_service(
+        &mut self,
+        id: &str,
+    ) -> &mut PersistentIncomingServiceState {
+        match self
+            .incoming_services
+            .iter_mut()
+            .position(|service| service.invitation_id == id)
+        {
+            // we have to use index, see https://github.com/rust-lang/rust/issues/21906
+            Some(index) => &mut self.incoming_services[index],
+            None => {
+                self.incoming_services.push(PersistentIncomingServiceState {
+                    invitation_id: id.to_string(),
+                    enabled: true,
+                    name: None,
+                });
+                self.incoming_services.last_mut().unwrap()
+            }
+        }
+    }
+}
+
+impl AppState {
+    pub async fn load_services_from_invites(
+        &self,
+        accepted_invitations: Vec<InvitationWithAccess>,
+    ) {
+        let incoming_services_arc = self.incoming_services();
+        let mut guard = incoming_services_arc.write().await;
+        for invite in accepted_invitations {
+            // during the synchronization we only need to add new ones
+            if guard.find_by_id(&invite.invitation.id).is_some() {
+                continue;
+            }
+
+            if let Some(service_access_details) = invite.service_access_details {
+                let mut enabled = true;
+                let mut name = None;
+
+                if let Some(state) = self
+                    .model(|m| {
+                        m.incoming_services
+                            .iter()
+                            .find(|s| s.invitation_id == invite.invitation.id)
+                            .cloned()
+                    })
+                    .await
+                {
+                    enabled = state.enabled;
+                    name = state.name.clone();
+                }
+
+                let name = if let Some(name) = name {
+                    name
+                } else {
+                    match service_access_details.service_name() {
+                        Ok(name) => name,
+                        Err(err) => {
+                            warn!(%err, "Failed to get service name from access details");
+                            continue;
+                        }
+                    }
+                };
+
+                guard.services.push(IncomingService::new(
+                    invite.invitation.id,
+                    name,
+                    None,
+                    enabled,
+                    service_access_details,
+                ));
+            } else {
+                warn!(
+                    "No service access details found for accepted invitations {}",
+                    invite.invitation.id
+                );
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct IncomingService {
```

Comment:
A bit of description would be nice. In particular to explain that:

 - `id` and `access_details` come from the invitation
 - `name` and `enabled` can be changed by the user of the service
 - `port` is set when the service is running

Those 3 groups have very different lifecycles.

Reply:
Also the information that what makes each service unique is its service id, and that comes from the invitation id.